### PR TITLE
[Benchmark] Adding minimal, reproducible AWS S3 benchmark

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #847

### CODE REVIEW

**Review by ejguan — APPROVED** [review]
LGTM. I think it's good!

**Comment by kiukchung** [comment]
Bringing this back to life from an offline conversation I had with @msaroufim.

It would be interesting to see run this benchmark comparing fsspec-s3 vs the s3-plugin (the one that uses the pybinded s3 crt client).

My team at Amazon uses the s3-plugin (through torchdata) to stream-load json (text) training data from S3. We have millions of shards (file in s3) and each shard is between 1 ~ 10GB (single digit gigabytes) big. Using torchdata's builtin s3 datapipe (e.g. `load_files_from_s3()`) I'm getting a throughput of 1GiB/sec out of the box from a single dataloader worker. 

While the example benchmark here is great, it falls a bit short when it comes to real-life scenarios:

1. For serious use-cases, the size of the files in S3 are in the orders of single-digit GB rather than MB
2. For non-image (binary) examples, no prefetching is required since one can stream-read from S3 into StringIOBuffer. This is preferable to prefetching large S3 files which has drawbacks like needing to provision instances with larger instance volumes, and making sure to clean up the prefetches when the job finishes/fails.

While the fsspec-s3 reader (which relies on aioboto3 for asyncio reads) is great for composing a batch from multiple smaller shards, the s3 c++ pybind client would be much more performant when one cannot take advantage of asyncio parallelism when loading few large shards.

It would be a shame to ditch support for the s3-plugin in favor of a more pythonic fsspec-s3 as hinted in these issues:

1. https://github.com/pytorch/data/issues/871
2. https://github.com/pytorch/data/issues/802

I understand the portability/maintainability concern of shipping c-bindings with python packages. Happy to discuss solutions/collaborations from our end to ensure that torchdata users have options for both s3 datapipes.

**Comment by ejguan** [comment]
Thanks for providing this valuable feedback. I appreciate that you have shown us different use cases loading from S3. We previously rely on Vision examples to run the benchmarking. 

> I understand the portability/maintainability concern of shipping c-bindings with python packages

This is generally not a problem as we have already shipped this packages with S3 plugin enabled. The main reason we currently are pivoting users to use `fsspec`-based DataPipe is a concern raised previously. See: https://github.com/pytorch/data/issues/500

Without clear understanding on what causes S3-plugin having worse performance in either small files or big archives, it's hard for us to recommend users to rely on S3-plugin. And, we are kind short-of-hand on investigating the root cause or best practice on loading from S3. (I have to admit that the dataset with large archive might still be smaller than GB-scale.)

**Comment by NivekT** [comment]
Hi @kiukchung,

Thanks for looking into this!

> It would be interesting to see run this benchmark comparing fsspec-s3 vs the s3-plugin (the one that uses the pybinded s3 crt client).

In our Slack channel "aws_s3_plugin", I attached a file showing the benchmark results comparing in two. The setup is very similar to the one in this PR. Let us know your thoughts on that.

I agree that the two points that you mentioned would be interesting to look at, but as @ejguan mentioned, we are a bit short staffed at the moment. If you have publicly reproducible results/investigations, we can definitely look at it.

**Comment by kiukchung** [comment]
Thanks @NivekT  and @ejguan. I spent the day running some benchmarks between s3io vs fsspec on our datasets in s3. Will comment on https://github.com/pytorch/data/issues/500.
